File: src/main/java/com/example/the_machine/repo/RunRepository.java
package com.example.the_machine.repo;

import com.example.the_machine.db.RunEntity;
import java.util.Optional;
import java.util.UUID;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.stereotype.Repository;

@Repository
public interface RunRepository extends JpaRepository<RunEntity, UUID> {

  Optional<RunEntity> findTopByProjectIdOrderByStartedAtDesc(UUID projectId);
}

File: src/main/kotlin/com/example/the_machine/repo/RunRepository.kt
package com.example.the_machine.repo

import com.example.the_machine.domain.RunEntity
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.stereotype.Repository
import java.util.*

@Repository
interface RunRepository : JpaRepository<RunEntity, UUID> {

  fun findTopByThreadIdOrderByStartedAtDesc(threadId: UUID): RunEntity?
}